File: python/0347.py
from typing import List
from collections import defaultdict

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        num_to_count = defaultdict(int)
        count_to_nums = [[] for i in range(len(nums) + 1)]
        for num in nums:
            num_to_count[num] += 1
        for num, count in num_to_count.items():
            count_to_nums[count].append(num)
        top_k = []
        for idx in range(len(count_to_nums) - 1, 0, -1):
            for num in count_to_nums[idx]:
                top_k.append(num)
                if len(top_k) == k:
                    return top_k



                

File: go/0347.go
package leetcode

func topKFrequent(nums []int, k int) []int {
	numFreqs := make(map[int]byte, len(nums))
	for _, num := range nums {
		numFreqs[num]++
	}
	buckets := make([][]int, len(nums)+1)
	for num, freq := range numFreqs {
		buckets[freq] = append(buckets[freq], num)
	}
	top_k := make([]int, 0, k)
	for i := len(buckets) - 1; i > 0 && k > 0; i-- {
		if len(buckets[i]) > 0 {
			if len(buckets[i]) > k {
				top_k = append(top_k, buckets[i][:k]...)
				k = 0
			} else {
				top_k = append(top_k, buckets[i]...)
				k -= len(buckets[i])
			}
		}
	}
	return top_k
}
